feat(core): add bun package manager

Jordan-Hall opened the pull request:
Bun uses yarn lock for it's binary file. Running the binary will produce the content of a yarn lock file (v1)

Other option is to use the -y command on add and install. This will create a yarn lock file and then createLockFile can just modify the yarn.lock file instead?

This is the PR made from #19113 and pushed due to #22402 being closed.

PS Bun feels more stable since the PR was first created!

This PR will resolve #22283 and start of #21075

Jordan-Hall commented on packages/nx/src/plugins/js/lock-file/lock-file.ts:
```diff
@@ -191,6 +213,12 @@ export function createLockFile(
+    if (packageManager === 'bun') {
+      output.log({
+        title:
+          "Unable to create bun lock files. Run bun install it's just as quick",
```
@edbzn  would it be worth calling stringifyYarnLockfile?

Jordan-Hall replied:
My thinking is moving the following into a util

```
const lockFile = createLockFile(
      builtPackageJson,
      context.projectGraph,
      packageManager
    );
    writeFileSync(
      `${outDirRelativeToWorkspaceRoot}/${getLockFileName(packageManager)}`,
      lockFile,
      {
        encoding: 'utf-8',
      }
    );
    ```
    
    its duplicated code but run IF check for bun and use the bun CLI to generate the binary file

edbzn replied:
Worth asking @meeroslav, he knows best how to deal with lockfiles

meeroslav replied:
Since there doesn't seem to be a way to generate `bun.lockb` without running the install, I would use the `stringifyYarnLockfile` to be as close as possible to the intended package versions.

Jordan-Hall replied:
@jaysoo  I've taken another look at this and asked in Bun if we can manually create the lock file but we cant :/ Reading a lock file is absolutely fine packages/nx/src/plugins/js/index.ts. The issue here is Bun lock file is a binary file. Only way to do it would be to execute Bun CLI to create the lock file on bun install

meeroslav replied:
Users won't be able to run `bun install --frozen-lockfile` since there would only be `yarn.lock`.

jaysoo replied:
👍 for creating util to execute bun CLI and returning the yarn lock file contents as string. The result can then be passed to `stringifyYarnLockFile` since it should compatible (just check for `# bun` and `version "workspace:..."` strings. We may or may not need to sanitize.

Jordan-Hall replied:
Just to confirm, you wanting to create a yarn.lock file here (with the additional bun hash?)

meeroslav replied:
We can't create bun hash since we can't generate `bun.lockb` to calculate hash off.

But by having a `yarn.lock` file, when running `bun install` it should use versions of the packages specified in `yarn.lock` to construct new bun lock file. At least that is what I'm hoping for.

Can you verify this?

Basically - if you take a yarn lock file with outdated nested dependencies (one of the fixture ones would work fine), then run bun install and finally bun bun.lockb - the result should be the same versions of packages.

meeroslav replied:
That's the best we can do right now I'm afraid

jaysoo replied:
We can also do the lockfile generation (as `yarn.lock`) in a follow-up PR. I don't think we need to handle it here.

Jordan-Hall replied:
I'm looking at generating the binary file. Happy to update snapshot so this can go in and work on that after? Also love to get involved in adding runtime and bun test. I have plugin that does it for now, but be nice if NX can add native support

Jordan-Hall replied:
@jaysoo  If you want to make the yarn.lock a follow up then I think this PR is completed?

laynef replied:
I reused Yarn in my PR bro. They still had problems with it. The people want Bun support! 

If it builds and runs, then that's a start.

Help us brothers out for taking the time to do this for you.

It was my birthday btw when I built a working version for your project.

Jordan-Hall replied:
This is different, this is to handle the manual creation of lock files not the reading of the lock file. Nx team doing a great job BTW, they been so many unstable issues. The chunk of this was taken from my PR back in Sep 11, 2023. But I closed it due to Bun having too many bugs

Jordan-Hall replied:
@laynef  you can always use nx-bun to get started for now

Jordan-Hall commented:
cc: @JamesHenry

Jordan-Hall commented:
I was working in WSL. I'll fix the test in the morning on my Linux laptop

Jordan-Hall commented on packages/nx/src/utils/package-manager.spec.ts:
```diff
@@ -58,6 +62,27 @@ describe('package-manager', () => {
+      });
+      const packageManager = detectPackageManager();
+      expect(packageManager).toEqual('bun');
+      expect(fs.existsSync).toHaveBeenCalledTimes(3);
```
Can someone explain to me how this could possible be 6 like NPM 1 :/ The two failing tests are not making sense

JamesHenry commented on packages/nx/src/plugins/js/index.ts:
```diff
@@ -51,7 +52,10 @@ export const createNodes: CreateNodes = [
+    const lockFileContents =
+      packageManager !== 'bun'
+        ? readFileSync(lockFilePath).toString()
+        : execSync(lockFilePath).toString();
```
Shouldn't it be `bun bun.lockb` as the command?

Jordan-Hall replied:
Addressed in the other comment

JamesHenry commented on packages/nx/src/plugins/js/index.ts:
```diff
@@ -91,7 +95,10 @@ export const createDependencies: CreateDependencies = (
+    const lockFileContents =
+      packageManager !== 'bun'
+        ? readFileSync(lockFilePath).toString()
+        : execSync(lockFilePath).toString();
```
Shouldn't it be `bun bun.lockb` as the command?

Jordan-Hall replied:
It's checking on the actual package manager and not the lock file.  I thought it would be more readable to check the package manager saying we already have the reference. Otherwise its checking the path

meeroslav replied:
Running just `bun.lockb` throws an error. I'm not sure I understand the comment above, but I believe we still need to parse it with `bun bun.lockb` as @JamesHenry suggested to be able to extract the hash (or hash it altogether) unless there is another way to get that hash out of the lockfile.

Jordan-Hall replied:
bun.lockb is self executing (at least on linux) Happy to change this to bun bun.lockb as it makes more sense

meeroslav replied:
It doesn't seem self-executing on mac OS, so please use the `bun bun.lockb` as that should work across the OSs.

JamesHenry reviewed (dismissed):
Thanks a lot for this @Jordan-Hall!

We will need some e2e coverage in addition to the comments

Jordan-Hall commented on e2e/utils/get-env-info.ts:
```diff
@@ -28,6 +28,8 @@ export function detectPackageManager(dir: string = ''): PackageManager {
     : existsSync(join(dir, 'pnpm-lock.yaml')) ||
       existsSync(join(dir, 'pnpm-workspace.yaml'))
     ? 'pnpm'
+    : existsSync(join(dir, 'bun.lockb'))
```
Already changed. I

Jordan-Hall commented on packages/nx/src/utils/package-manager.ts:
```diff
@@ -343,12 +368,16 @@ export async function packageRegistryView(
+     *
+     * Bun has a pm ls function but it only relates to it's lockfile
+     * and acts differently from all other package managers
+     * from Jarred"it probably would be bun pm view <package-name>"
```
@JamesHenry  Not that I'm aware of . I cant find anything in their repo. It was a commend he made in discord https://discord.com/channels/876711213126520882/1152139091484344330/1152141896093794315

Jordan-Hall commented on e2e/workspace-create/src/create-nx-workspace.test.ts:
```diff
@@ -414,7 +416,8 @@ describe('create-nx-workspace', () => {
+          packageManagerLockFile['npm'],
+          packageManagerLockFile['bun']
         );
         process.env.SELECTED_PM = packageManager;
```
Cheers, I've added a test around this

Jordan-Hall commented on packages/nx/src/utils/package-manager.spec.ts:
```diff
@@ -10,6 +10,9 @@ import {
 describe('package-manager', () => {
   describe('detectPackageManager', () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
```
Without this it appears these tests are not logically because it increase the call and depends on the state. Doing this allows you to predict correctly whats changing.

Jordan-Hall commented:
@JamesHenry I think this is all good now. The failed test is due to snapshot changes. Are you happy with me just updating the snapshot?

meeroslav commented on packages/nx/src/plugins/js/index.ts:
```diff
@@ -51,7 +52,10 @@ export const createNodes: CreateNodes = [
+    const lockFileContents =
+      packageManager !== 'bun'
+        ? readFileSync(lockFilePath).toString()
+        : execSync('bun bun.lockb').toString();
```
```suggestion
        : execSync(`bun ${lockFilePath}`).toString();
```

meeroslav replied:
Please apply the same fix on line 101.

Jordan-Hall replied:
Done that. Sorry for the confusion; i only knew about self-excuting on linux

meeroslav replied:
Optionally we can use ``execSync(`bun ${lockFilePath}`, { encoding: 'utf8' })`` to ensure proper decoding while stringification.

meeroslav reviewed (approved):
Looks good. As @jaysoo mentioned, we can add pruning logic subsequently.

meeroslav commented:
> I am looking at the pruning. Never looked at zig code but it wonder if porting this would help https://github.com/oven-sh/bun/blob/main/src/install/lockfile.zig

Unless they have an npm package that exposes generator as a Node function, we don't want to "reimplement" it within Nx. For the time being, we can use yarn lock file generation and once there is a way for `bun` to generate a lock file without running the install we'll switch to that.

Something like `bun install --lock-file-only` or `bun install --prune` would be perfect.

Jordan-Hall commented on packages/workspace/src/generators/new/schema.json:
```diff
@@ -56,7 +56,7 @@
       "description": "The package manager used to install dependencies.",
       "type": "string",
-      "enum": ["npm", "yarn", "pnpm"]
+      "enum": ["npm", "yarn", "pnpm", "bun"]
```
Nice spot! Cent believe I missed that.

jaysoo replied:
All good. I tested it out locally and it's working now. Checking a few more things.

Jordan-Hall commented on packages/nx/src/utils/package-manager.ts:
```diff
@@ -142,6 +144,20 @@ export function getPackageManagerCommand(
+        add: 'bun install',
+        addDev: 'bun install -D',
+        rm: 'bun rm',
+        exec: 'bun',
```
I wasn't sure whether to use the --bun flag to force inclusion in bun. However, although bun isn't exactly a 1-to-1 match yet, I thought it could introduce too many non-issues to NX.

Maybe 1 to follow up with eventually

jaysoo replied:
@Jordan-Hall Where would the `--bun` flag go? It should be consistent with other package managers.

Jordan-Hall replied:
exec: 'bun --bun'
dlx: 'bunx --bun',
        run: (script: string, args: string) => `bun run --bun ${script} -- ${args}`,


It basically stops node shebang  and forces it to bun

jaysoo commented:
One thing I noticed when testing locally is that setting local registry in `$HOME/.bunfig.toml` didn't always work for me unless I run `bun install --no-cache`. Not sure what the issue it, but perhaps it's related to https://github.com/oven-sh/bun/issues/4901.

This would affect e2e testing.

jaysoo commented:
One thing I noticed when testing locally is that setting local registry in `$HOME/.bunfig.toml` didn't always work for me unless I run `bun install --no-cache`. Not sure what the issue it, but perhaps it's related to https://github.com/oven-sh/bun/issues/4901.

This would affect e2e testing.

Jordan-Hall commented:
@jaysoo one thing I haven't done is update verdaccio to edit bunfig.toml to add a custom registery

Jordan-Hall commented:
> One thing I noticed when testing locally is that setting local registry in `$HOME/.bunfig.toml` didn't always work for me unless I run `bun install --no-cache`. Not sure what the issue it, but perhaps it's related to [oven-sh/bun#4901](https://github.com/oven-sh/bun/issues/4901).
> 
> This would affect e2e testing because we need to set it to:
> 
> ```
> // $HOME/.bunfig.toml
> [install]
> registry = "http://localhost:4873"
> ```

I believe this is by design. When you say it wasn't working what do you mean? By default bun always go to cache as long as the major version is matching

jaysoo commented:
> @jaysoo one thing I haven't done is update verdaccio to edit bunfig.toml to add a custom registery

We should do this here or as a follow-up to add bun to nightly tests, since it will be needed there.

> I believe this is by design. When you say it wasn't working what do you mean? By default bun always go to cache as long as the major version is matching

I would publish a new version (e.g. `19.0.0`) locally, update `.bunfig.toml`, then run `bun install`. The install fails due to `19.0.0` not being on npm registry, so I have to pass `--no-cache` to make it work. Seems like a bug, but it's nothing to do with Nx, we can just always run with `--no-cache` for e2e tests.

Jordan-Hall commented:
I'll update the test shortly. Is they anything else we need for this or want me to look at?

jaysoo commented:
> I'll update the test shortly. Is they anything else we need for this or want me to look at?

I think we're good. When we enable nightlies we'll make sure all the local registry support is there as well.

Jordan-Hall commented:
I'm happy adding that now if you show me where

meeroslav commented:
> I'm happy adding that now if you show me where
@Jordan-Hall not need, we'll do that in a separate PR. Let's get this one over the line

Jordan-Hall commented:
@meeroslav any chance you can update that last remaining snapshot for me. Sorry on windows and snapshot changes too much compared to doing it on linux. Not got my linux laptop on me sorry

Jordan-Hall commented:
So latest pr revase broke it again -_- let me know when you want this in / ready and I'll fix it again.

jaysoo commented:
> So latest pr revase broke it again -_- let me know when you want this in / ready and I'll fix it again.

Looks like it's green now. I'll do more testing on it, it looked good during my last test. If it's good, we can probably get it in for Nx 19.

Jordan-Hall commented:
> > So latest pr revase broke it again -_- let me know when you want this in / ready and I'll fix it again.
> 
> Looks like it's green now. I'll do more testing on it, it looked good during my last test. If it's good, we can probably get it in for Nx 19.

Thank you. Let me know if you find any issues. Next stop bun executors 😅

enaluz commented:
Is there an ETA on when this can be merged in and be usable? I'm starting a new React/React Native project for my company this week or next and I really want to use Bun + Nx 👀

mehrangta commented:
> Is there an ETA on when this can be merged in and be usable? I'm starting a new React/React Native project for my company this week or next and I really want to use Bun + Nx 👀
> 
> If I decided to go with an alternative package manager just to get the project started (like `pnpm` for example), would it involve significant overhead to switch over to Bun?

As far as I know, bun cannot change pnpm lockfile to bun lockfile. so I would choose classic npm to start with. (also workspaces in pnpm are a little bit different too)

jaysoo commented:
Pushed one more fix and doing more testing. We're ready to merge Nx 19 features, so we should be merging this very soon.

jaysoo commented:
Quick update on this PR, we are planning to include this in Nx 19.1 rather than 19. The reason is that Nx 19 is a stability release to address some issues around Project Crystal, so we're pushing some new features to Nx 19.1, including Bun.

ItsWendell commented:
@jaysoo any update or indication _when_ this will land in 19.1 (beta)?

jaysoo commented:
> @jaysoo any update or indication _when_ this will land in 19.1 (beta)?

I think we can merge it soon once green again.

benpsnyder commented:
@Jordan-Hall please complete https://github.com/Jordan-Hall/nx/pull/2/

Jordan-Hall commented:
> @Jordan-Hall please complete [Jordan-Hall#2](https://github.com/Jordan-Hall/nx/pull/2)

I'll sort it later. I was going to fix this tonight when little one asleep. Sorry didn't get any notification

Jordan-Hall commented:
> @Jordan-Hall please complete [Jordan-Hall#2](https://github.com/Jordan-Hall/nx/pull/2)

I think i  want @jaysoo  view on this. Now that you have added getRegistryUrl. How do you want to do this in bun as it doesnt have the same support

Jordan-Hall commented:
I cleaned up the commits. I've added a comment where getRegistryUrl is also being used internally, so it doesn't waste time trying to run the command.
Hopefully this can no be merged in over a year after it was created 😂

Jordan-Hall commented:
@jaysoo  could you let me know why its failing please https://app.circleci.com/pipelines/github/nrwl/nx/50899/workflows/5f623f71-d6a0-47c9-b41b-3d3bf7845f5a/jobs/423702 its looking like code owner if im not mistaken?

loicnestler commented:
What's the status on this? 👀

Jordan-Hall commented:
> > What's the status on this? 👀
> 
> Paying Nx Cloud customer/partner here ... monthly spend is probably $1500/month across our internal and clients we directed to Nx Who do I have to wire transfer to in order to get this PR to close? And what amount? LOL. Sorry. Don't mean to spam. Just hitting refresh on my browser on this PR every 2 hours 😊 @jaysoo 👀 👀

I really hate people using that as a forced. I think it's my fault its not building I'll check tonight to see if it's not codeowner. Please show some respect though.

benpsnyder commented:
> > > What's the status on this? 👀
> > 
> > 
> > Paying Nx Cloud customer/partner here ... monthly spend is probably $1500/month across our internal and clients we directed to Nx Who do I have to wire transfer to in order to get this PR to close? And what amount? LOL. Sorry. Don't mean to spam. Just hitting refresh on my browser on this PR every 2 hours 😊 @jaysoo 👀 👀
> 
> I really hate people using that as a forced. I think it's my fault its not building I'll check tonight to see if it's not codeowner. Please show some respect though.

I literally mean I am willing to contribute more financially to prioritize. Not trying to force anyone anything. Just willing to pay for it to be prioritized. Thank you all for your efforts! Send me a DM for some $$ if there is an easy link and I will remit.

Everybody love everybody :D

Jordan-Hall commented:
Money worthless to me. Please think before posting. I've had this as a oat h file and bun plugin for a while before this PR. DO BETTER.

In meantime I'm home soon to check why it's failing !!!

jaysoo commented:
@benpsnyder As mentioned before, we'll be releasing this with Nx 19.1, which is wrapping up development by end of next week (May 31st). We'll get this merge and ready to go before then. The actual release will likely be the first week of June.

@Jordan-Hall If you need help with CI and failing checks let me know, we'll help you get this pushed through.

Jordan-Hall commented:
@jaysoo  are you on slack or discord. I'm confused why this is failing. Build and lint ok. Looks to me its a codeowner issue

FrozenPandaz commented on packages/workspace/src/generators/ci-workflow/__snapshots__/ci-workflow.spec.ts.snap:
```diff
@@ -216,6 +216,249 @@ jobs:
+    # The "--stop-agents-after" is optional, but allows idle agents to shut down once the "build" targets have been requested
+    # - npx nx-cloud start-ci-run --distribute-on="5 linux-medium-js" --stop-agents-after="build"
+
+    - npm ci
```
These snapshots seem off. They use `npm`. Shouldn't they use `bun`?

Jordan-Hall replied:
Debugging this now. It appears its not detecting the bun.lockb

Jordan-Hall replied:
Resolved. Also changed all the template for the pipelines. Github doesn't require the cache because oven-sh/setup-bun@v1 comes with cache support. The others copy pnpm and install bun via npm https://www.npmjs.com/package/bun

FrozenPandaz commented on packages/create-nx-workspace/src/utils/package-manager.ts:
```diff
@@ -79,6 +81,12 @@ export function getPackageManagerCommand(
+      return {
+        install: 'bun install --silent --ignore-scripts',
+        exec: 'bunx',
+        globalAdd: 'bun install -g',
```
Could you please add a comment here and elsewhere where we define the `bun` PackageManager linking to the issue which tracks the `config` command please?

Jordan-Hall replied:
Are you on about the issue relating for the getRegistryUrl?  I'll see if there is an issue for it.

FrozenPandaz reviewed (changes requested):
Thank you for your work here! :+1:

benpsnyder commented:
![image](https://github.com/nrwl/nx/assets/707567/86f68e13-5e06-4354-8a76-d7da5d056dd8)

We are setting aside resources in the coming weeks to battle test bun in our repos with Nx and we'll submit PRs and open issues as necessary. Very excited about this. Thanks to all

Jordan-Hall commented:
I've had bun test working with NX it really good. Happy to help get a plugin in NX to support both runtime and tests. Would need guidance though by the team. I feel the node should turn into runtime and work with node and bun at least the generators

Jordan-Hall commented:
@FrozenPandaz  I think this is ready now. Let me know if you have any queries or any more issues :)

jaysoo commented:
Trying a run here to see if it'll go green https://github.com/nrwl/nx/pull/25066

Need to debug further on why CI is failing.

Jordan-Hall commented:
> Trying a run here to see if it'll go green #25066
> 
> Need to debug further on why CI is failing.

Thank you. I'm around now on laptop if you need anything

jaysoo commented:
Trying a run here to see if it'll go green https://github.com/nrwl/nx/pull/25066

Need to debug further on why CI is failing.

Jordan-Hall commented:
Yes yes yes @jaysoo merge before it goes again 🤣

jaysoo commented:
Pushed a WIP PR to see how e2e tests are when we force Bun, since they are running with pnpm right now. If things are green or mostly green then we're good to go.

Jordan-Hall commented:
Can I ask what the issue was with the ci :/ be good for future reference if I make anymore prs

jaysoo commented:
We'll need a follow-up to make sure that `--generatePackageJson` for executors that support it won't error out due to missing lockfiles. For Bun, we'll just skip the lockfile for now, and discuss whether generating `yarn.lock` file would be preferable. 

This is the code we need to fix: https://github.com/nrwl/nx/blob/master/packages/js/src/utils/package-json/update-package-json.ts#L106-L118.

Jordan-Hall commented:
> We'll need a follow-up to make sure that `--generatePackageJson` for executors that support it won't error out due to missing lockfiles. For Bun, we'll just skip the lockfile for now, and discuss whether generating `yarn.lock` file would be preferable.
> 
> This is the code we need to fix: https://github.com/nrwl/nx/blob/master/packages/js/src/utils/package-json/update-package-json.ts#L106-L118.

Happy to make that PR later

jaysoo commented:
Here's the PR :)

https://github.com/nrwl/nx/pull/25158

jaysoo commented:
You can try this now in canary.

```
npx create-nx-workspace@canary --packageManager=bun
```

benpsnyder commented:
First issue filed. Going to spend some hours debugging this and will post findings there.

`When using Bun and building within Dockerfile - Failed to process project graph. Run "nx reset" to fix this.' re: 'nx/js/dependencies-and-lockfile` (https://github.com/nrwl/nx/issues/25965)
